### PR TITLE
SQL Server input plugin - Enable Azure Active Directory (AAD) authentication support 

### PR DESCRIPTION
**Pull request by avinash-nigam**
### Required for all PRs:

- [ ] Associated README.md updated.
- [ ] Has appropriate unit tests.

Associated to feature request - [Azure Active Directory (AAD) authentication support in SQL Server input plugin](https://github.com/influxdata/telegraf/issues/8808#issue-801695311)

### CODE REVIEW

File: plugins/inputs/sqlserver/sqlserver.go [hunk 1]
```diff
@@ -244,11 +247,36 @@ func (s *SQLServer) Gather(acc telegraf.Accumulator) error {
...
 
 func (s *SQLServer) gatherServer(server string, query Query, acc telegraf.Accumulator) error {
-	// deferred opening
-	conn, err := sql.Open("mssql", server)
-	if err != nil {
-		return err
+	// deferred connection opening
+	var conn *sql.DB
+
+	if strings.Contains(strings.ToLower(server), strings.ToLower("password=")) {
```
Comment (srebhan) [inline plugins/inputs/sqlserver/sqlserver.go:253]: I rather like to see a regular expression that makes sure that `password` is the beginning of the parameter and not e.g. part of `o_i_lost_my_password=damn` or `ignorepassword=true`. :-)

File: go.mod [hunk 2]
```diff
@@ -12,6 +12,7 @@ require (
 	github.com/Azure/azure-storage-queue-go v0.0.0-20181215014128-6ed74e755687
 	github.com/Azure/go-ansiterm v0.0.0-20170929234023-d6e3b3328b78 // indirect
 	github.com/Azure/go-autorest/autorest v0.9.3
+	github.com/Azure/go-autorest/autorest/adal v0.8.1
```
Comment (srebhan) [inline go.mod:15]: Please also add this to `docs/LICENSE_OF_DEPENDENCIES.md`. Currently it seems like our check-script is on strike, so watch the output with care. :-)

File: plugins/inputs/sqlserver/sqlserver.go [hunk 3]
```diff
@@ -244,11 +247,36 @@ func (s *SQLServer) Gather(acc telegraf.Accumulator) error {
...
+		var err error
+		conn, err = sql.Open("mssql", server)
+
+		if err != nil {
+			return err
+		}
+	} else {
+		// when password is not provided in connection string, use AAD auth with an MSI token
+
+		tokenProvider, err := getMSITokenProvider()
```
Comment (masree) [inline plugins/inputs/sqlserver/sqlserver.go:265]: Do we need to get the token for all the servers or can we get the token once and leverage it for all servers since t is from the same VM?

**Comment by masree** [comment]
I think we should update SQL Server plugin doc to indicate AAD support using MSI

File: plugins/inputs/sqlserver/sqlserver.go (hunk 3 above)
Comment (masree) [inline plugins/inputs/sqlserver/sqlserver.go:265]: I believe for all servers under a single VM, token will be same and we can re-use. Will fetching of token happen once per VM or is this once per server?  If it is latter, is it possible to fetch the token once and re-use for all servers in one VM?

File: plugins/inputs/sqlserver/sqlserver.go [hunk 4]
```diff
@@ -242,11 +245,36 @@ func (s *SQLServer) Gather(acc telegraf.Accumulator) error {
 }
 
 func (s *SQLServer) gatherServer(server string, query Query, acc telegraf.Accumulator) error {
-	// deferred opening
-	conn, err := sql.Open("mssql", server)
-	if err != nil {
-		return err
+	// deferred connection opening
```
Comment (masree) [inline plugins/inputs/sqlserver/sqlserver.go:248]: nit: this comment should go below when opening the connection

File: plugins/inputs/sqlserver/sqlserver.go [hunk 5]
```diff
@@ -242,11 +245,36 @@ func (s *SQLServer) Gather(acc telegraf.Accumulator) error {
...
+	var conn *sql.DB
+
+	if strings.Contains(strings.ToLower(server), strings.ToLower("password=")) {
+		// when password is provided in connection string, use SQL auth
+
```
Comment (masree) [inline plugins/inputs/sqlserver/sqlserver.go:253]: Can we log the database and server saying SQL Auth vs AAD auth? That would help in troubleshooting when something fails

File: plugins/inputs/sqlserver/sqlserver.go [hunk 6]
```diff
@@ -242,11 +245,36 @@ func (s *SQLServer) Gather(acc telegraf.Accumulator) error {
...
+		var err error
+		conn, err = sql.Open("mssql", server)
+
+		if err != nil {
+			return err
+		}
+	} else {
+		// when password is not provided in connection string, use AAD auth with an MSI token
+
```
Comment (masree) [inline plugins/inputs/sqlserver/sqlserver.go:262]: Add logging for AAD auth maybe?

File: plugins/inputs/sqlserver/sqlserver.go [hunk 7]
```diff
@@ -242,11 +245,36 @@ func (s *SQLServer) Gather(acc telegraf.Accumulator) error {
...
+		tokenProvider, err := getMSITokenProvider()
```
Comment (masree) [inline plugins/inputs/sqlserver/sqlserver.go:263]: For SQL VM, should we throw an exception and fail sooner here? SQL VM does not support AAD auth

File: plugins/inputs/sqlserver/sqlserver.go (hunk 1 above)
Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserver.go]: Addressed the same using regular expression

**Comment by avinash-nigam** [comment]
> I think we should update SQL Server plugin doc (readme) to indicate AAD support using MSI. Also indicate how an AAD login is created on SQL side using TSQL

Added content to readme file about its usage and instructions

File: plugins/inputs/sqlserver/sqlserver.go (hunk 4 above)
Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserver.go]: fixed

File: plugins/inputs/sqlserver/sqlserver.go (hunk 7 above)
Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserver.go:270]: Added a conditional check for DatabaseType = SQLServer i.e. SQL VM to error out if AAD auth is used

File: plugins/inputs/sqlserver/sqlserver.go (hunk 5 above)
Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserver.go:255]: Yes, I think it'd be useful to have the Server/Database name and Auth mode details while troubleshooting. Once the above mentioned PR is merged, we can extend the metrics to capture this.

File: plugins/inputs/sqlserver/sqlserver.go [hunk 8]
```diff
@@ -275,6 +282,7 @@ func (s *SQLServer) gatherServer(server string, query Query, acc telegraf.Accumu
 		conn = sql.OpenDB(connector)
 	}
 
+	// deferred connection closure
```
Comment (srebhan) [inline plugins/inputs/sqlserver/sqlserver.go:285]: Please remove as it does not add any information.

File: plugins/inputs/sqlserver/sqlserver.go [hunk 9]
```diff
@@ -260,6 +262,11 @@ func (s *SQLServer) gatherServer(server string, query Query, acc telegraf.Accumu
 	} else {
 		// when password is not provided in connection string, use AAD auth with an MSI token
 
+		// return error if AAD Auth is used for DatabaseType SQLServer since AAD Auth is only supported for Azure SQL Database or Azure SQL Managed Instance
```
Comment (srebhan) [inline plugins/inputs/sqlserver/sqlserver.go:265]: Please shorten this to `AAD Auth is only supported for Azure SQL Database or Azure SQL Managed Instance`.

File: plugins/inputs/sqlserver/sqlserver.go [hunk 10]
```diff
@@ -245,10 +245,12 @@ func (s *SQLServer) Gather(acc telegraf.Accumulator) error {
 }
 
 func (s *SQLServer) gatherServer(server string, query Query, acc telegraf.Accumulator) error {
-	// deferred connection opening
+	// connection object
 	var conn *sql.DB
 
-	if strings.Contains(strings.ToLower(server), strings.ToLower("password=")) {
+	// setup connection based on authentication
+	rx := regexp.MustCompile("(?i)\\b(?:;(Password=\\w+))\\b")
```
Comment (srebhan) [inline plugins/inputs/sqlserver/sqlserver.go:252]: Not sure but is `dsn="Password=verysecret"` also a valid DSN (without the semicolon)?

File: go.mod (hunk 2 above)
Comment (srebhan) [reply go.mod:15]: Ok seems like it is covered by `github.com/Azure/go-autorest/autorest` already. Please mark as resolved.

**Review by srebhan — COMMENTED** [review]
Please check my comments in the code. @masree any comments?

File: plugins/inputs/sqlserver/README.md [hunk 11]
```diff
@@ -166,9 +166,34 @@ GO
...
+        DROP USER [<Monitoring_VM_Name>]
+    END')
+EXECUTE ('CREATE USER [<Monitoring_VM_Name>] FROM EXTERNAL PROVIDER')
+EXECUTE ('GRANT VIEW DATABASE STATE TO [<Monitoring_VM_Name>]')
+```
+- On the SQL Server resource of the database(s) being monitored, go to "Firewalls and Virtual Networks" tab and allowlist the monitoring VM IP address.
+- On the Monitoring VM, update the telegraf config file with the database connection string in the following format.
+```toml
+  servers = [
+    "Server=<Azure_SQL_Server_Name>.database.windows.net;Port=1433;Database=<Azure_SQL_Database_Name>;Authentication=Active Directory Integrated;app name=telegraf;log=1;",
```
Comment (masree) [inline plugins/inputs/sqlserver/README.md:195]: Can we explicitly mention is supported for SQL MI and SQL DB? Also call out it is not supported for SQL server on a VM

**Comment by masree** [comment]
> Please check my comments in the code. @masree any comments?

Thank you for the review!

File: plugins/inputs/sqlserver/README.md [hunk 12]
```diff
@@ -172,9 +172,34 @@ GO
...
+
+Azure SQL Database supports 2 main methods of authentication: [SQL authentication and AAD authentication](https://docs.microsoft.com/en-us/azure/azure-sql/database/security-overview#authentication). The recommended practice is to [use AAD authentication when possible](https://docs.microsoft.com/en-us/azure/azure-sql/database/authentication-aad-overview).
 
+AAD is a more modern authentication protocol, allows for easier credential/role management, and can eliminate the need to include passwords in a connection string.
 
+To enable support for AAD authentication, we leverage the existing AAD authentication support in the [SQL Server driver for Go](https://github.com/denisenkom/go-mssqldb#azure-active-directory-authentication---preview)
 
+#### How to use AAD Auth with MSI
+
+- Configure managed identity for Azure resources on the Monitoring VM (the VM that'd connect to the SQL server/database) [using the Azure portal](https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/qs-configure-portal-windows-vm).
```
Comment (masree) [inline plugins/inputs/sqlserver/README.md:187]: would it work with both user and system identity?

Comment (avinash-nigam) [reply plugins/inputs/sqlserver/README.md:187]: User identity won't work here, since we retrieve token based on managed identity.

File: plugins/inputs/sqlserver/sqlserver.go (hunk 9 above)
Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserver.go]: done

File: plugins/inputs/sqlserver/sqlserver.go (hunk 8 above)
Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserver.go]: done

File: plugins/inputs/sqlserver/sqlserver.go (hunk 3 above)
Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserver.go]: Added in-memory token caching mechanism, it gets a token for the managed identity and use it until it finds it expired.

File: plugins/inputs/sqlserver/README.md (hunk 11 above)
Comment (avinash-nigam) [reply plugins/inputs/sqlserver/README.md:201]: Updated

File: plugins/inputs/sqlserver/README.md [hunk 13]
```diff
@@ -172,9 +172,34 @@ GO
...
+- On the database being monitored, create/update a USER with the name of the Monitoring VM as the principal using the below script. This might require allow-listing the client machine's IP address (from where the below SQL script is being run) on the SQL Server resource.
+```sql
+EXECUTE ('IF EXISTS(SELECT * FROM sys.database_principals WHERE name = ''<Monitoring_VM_Name>'')
```
Comment (masree) [inline plugins/inputs/sqlserver/README.md:190]: I would exclude the drop, and add only create user

File: plugins/inputs/sqlserver/sqlserver.go (hunk 10 above)
Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserver.go]: I've updated regex condition used for AAD auth applicability where I look for Authentication parameter instead of Password. However with regards to your concern about well-formed dsn or connection string, the intention of this check if the connection string provides a way to identify auth mechanism and not if its well-formed. We could check that but it should be part of validating in the input config file instead of this check alone.

Just for clarity, the new regex I'm using now "(?i)(^|[;])(?:(\s*Authentication=Active Directory Integrated\s*))([;]|$)" would take care of scenarios where the connection string contains the "Authentication=Active Directory Integrated" string anywhere (at the start or at the end or in the middle somewhere).

File: plugins/inputs/sqlserver/README.md (hunk 13 above)
Comment (avinash-nigam) [reply plugins/inputs/sqlserver/README.md:190]: The intention is to provide a "defensive" script, running it would not cause any harm. And the user always has the option to remove part of it if they deem it unnecessary.

File: plugins/inputs/sqlserver/sqlserver.go [hunk 14]
```diff
@@ -275,11 +286,41 @@ func (s *SQLServer) Gather(acc telegraf.Accumulator) error {
 }
 
 func (s *SQLServer) gatherServer(server string, query Query, acc telegraf.Accumulator) error {
-	// deferred opening
-	conn, err := sql.Open("mssql", server)
-	if err != nil {
-		return err
+	// connection object
+	var conn *sql.DB
```
Comment (masree) [inline plugins/inputs/sqlserver/sqlserver.go:290]: Heads up: Depending on which [PR ](https://github.com/influxdata/telegraf/pull/8596/files)gets merged to master, there  may be a possibility of merge conflict

Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserver.go:290]: thank you, yes I did notice that PR

File: plugins/inputs/sqlserver/README.md (hunk 12 above)
Comment (masree) [reply plugins/inputs/sqlserver/README.md:187]: User Identity is also a managed identity- https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-manage-ua-identity-portal#:~:text=1%20Sign%20in%20to%20the%20Azure%20portal%20using,of%20a%20user-assigned%20managed%20identity%20click%20its%20name. Can you help check? If it does not work, documentation might have to be explicit saying system-assigned identity.

Comment (avinash-nigam) [reply plugins/inputs/sqlserver/README.md:187]: Yes, this is not meant for a user-assigned managed identity, I'll update the verbiage to call out that only a system-assigned managed identity should be created for this case.

File: plugins/inputs/sqlserver/sqlserver.go [hunk 15]
```diff
@@ -275,11 +286,41 @@ func (s *SQLServer) Gather(acc telegraf.Accumulator) error {
...
+
+	// setup connection based on authentication
+	rx := regexp.MustCompile(`(?i)(^|[;])(?:(\s*Authentication=Active Directory Integrated\s*))([;]|$)`)
+
+	// when auth mechanism is mentioned as AAD in the connection string, use managed-identity token
+	if rx.MatchString(server) {
```
Comment (masree) [inline plugins/inputs/sqlserver/sqlserver.go:296]: Is "Authentication=Active Directory Integration" property a mandatory field in connection string? If that property is missing, and password is not specified, connection would still work right?

Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserver.go:296]: Yes, for AAD auth the connection string needs to include "Authentication=Active Directory Integration" property (I'm referring to [this ](https://docs.microsoft.com/en-us/azure/azure-sql/database/authentication-aad-configure?tabs=azure-powershell#active-directory-integrated-authentication-1) doc). With the current code, if the property is not included it'll be processed as SQL Auth, and for your example, it won't connect since there are no credentials.

Comment (masree) [reply plugins/inputs/sqlserver/sqlserver.go:296]: I think that is AAD and different from MSI auth. Can you help remove the property Authentication from your connection string and see if you're able to connect in your test systems. My understanding is that this is not a required field and connection should succeed. Please help verify.

Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserver.go:296]: Thank you @masree for highlighting and clarifying this. Yes, MSI auth does not require the Authentication property in the connection string. I have reverted the code to use existence of "Password" as the definition of whether to use SQL Auth or AAD/MSI Auth. Readme file has been correspondingly modified as well.

**Comment by masree** [comment]
How have you testing it? is there a testing plan you would like to link for us to take a look?
Example - 
Scenario1: Telegraf installed on VM and with connection strings using sqlauth only 
Scenario2: Telegraf installed on VM with connection strings using sqlauth and aadMSI (multiple connection strings) 
and so on

File: plugins/inputs/sqlserver/sqlserver.go [hunk 16]
```diff
@@ -408,6 +450,96 @@ func (s *SQLServer) Init() error {
 	return nil
 }
 
+// Get Token Provider by loading cached token or refreshed token
+func getTokenProvider() (func() (string, error), error) {
```
Comment (masree) [inline plugins/inputs/sqlserver/sqlserver.go:454]: Do you think it is better to move token related methods to a new class? That would help to keep this main class minimal.

File: plugins/inputs/sqlserver/sqlserver.go [hunk 17]
```diff
@@ -34,6 +36,12 @@ type Query struct {
 	OrderedColumns []string
 }
 
+// var to cache ADAL MSI token
+var adalToken *adal.Token
+
+// var for mutual exclusion lock
+var muCacheLock sync.RWMutex
+
```
Comment (srebhan) [inline plugins/inputs/sqlserver/sqlserver.go:44]: Is there a reason why you don't move this inside the struct? I assume this is the same for all `servers`!?

File: plugins/inputs/sqlserver/sqlserver.go [hunk 18]
```diff
@@ -272,15 +280,48 @@ func (s *SQLServer) Start(acc telegraf.Accumulator) error {
...
+			if err != nil {
+				acc.AddError(err)
+				return err
+			}
+		} else {
+			// otherwise assume AAD Auth with system-assigned managed identity (MSI)
+
+			// AAD Auth is only supported for Azure SQL Database or Azure SQL Managed Instance
+			if s.DatabaseType == "SQLServer" {
+				return fmt.Errorf("Database connection failed : AAD auth is not supported for SQL VM i.e. DatabaseType=SQLServer")
```
Comment (srebhan) [inline plugins/inputs/sqlserver/sqlserver.go:306]: Error texts have to start lowercase.

File: plugins/inputs/sqlserver/sqlserver.go [hunk 19]
```diff
@@ -272,15 +280,48 @@ func (s *SQLServer) Start(acc telegraf.Accumulator) error {
...
+			}
+
+			// get token from im-memory cache variable or from Azure Active Directory
+			tokenProvider, err := getTokenProvider()
+			if err != nil {
+				fmt.Printf("Error creating AAD token provider for system assigned Azure managed identity : %s", err.Error())
```
Comment (srebhan) [inline plugins/inputs/sqlserver/sqlserver.go:312]: Never `fmt.Printf`! Add a logger facility 
```
	Log telegraf.Logger `toml:"-"`
```
to the `SQLServer` struct and then use `s.Log.Warn()` or whatever to log the message. Alternatively you can just return the message as error message!?

File: plugins/inputs/sqlserver/sqlserver.go [hunk 20]
```diff
@@ -272,15 +280,48 @@ func (s *SQLServer) Start(acc telegraf.Accumulator) error {
...
+				acc.AddError(err)
+				return err
+			}
+
+			connector, err := mssql.NewAccessTokenConnector(serv, tokenProvider)
+			if err != nil {
+				fmt.Printf("Error creating the SQL connector : %s", err.Error())
```
Comment (srebhan) [inline plugins/inputs/sqlserver/sqlserver.go:319]: See above.

File: plugins/inputs/sqlserver/sqlserver.go [hunk 21]
```diff
@@ -423,6 +464,96 @@ func (s *SQLServer) Init() error {
...
+			}
+
+			// use the refreshed token
+			tokenString = spt.OAuthToken()
+		} else {
+			// use locally cached token
+			tokenString = token.OAuthToken()
+		}
+
+		muCacheLock.Unlock()
```
Comment (srebhan) [inline plugins/inputs/sqlserver/sqlserver.go:499]: Oh boy, if you return in between you will not release the lock!!!!!!!!!!!!!
Please use
```golang
muCacheLock.Lock()
defer muCacheLock.Unlock()
```
as a robust pattern wherever possible!

File: plugins/inputs/sqlserver/sqlserver.go [hunk 22]
```diff
@@ -423,6 +464,96 @@ func (s *SQLServer) Init() error {
...
+}
+
+// Load token from in-mem cache
+func loadToken() (*adal.Token, error) {
+	if adalToken == nil {
+		return nil, fmt.Errorf("Token is nil or Failed to load existing token")
+	}
+
+	return adalToken, nil
+}
```
Comment (srebhan) [inline plugins/inputs/sqlserver/sqlserver.go:518]: I don't get the intention behind this function... You could check for the token being `nil` on the caller side or am I missing something...

File: plugins/inputs/sqlserver/sqlserver.go [hunk 23]
```diff
@@ -423,6 +464,96 @@ func (s *SQLServer) Init() error {
...
+
+// Refresh token for the resource, and save to in-mem cache
+func refreshToken() (*adal.Token, error) {
+	// resource id for Azure SQL Database
+	const sqlAzureResourceID = "https://database.windows.net/"
```
Comment (srebhan) [inline plugins/inputs/sqlserver/sqlserver.go:523]: Should this be really fixed? And if so, is there a reason to put it here and not side-by-side with the other const declarations?

File: plugins/inputs/sqlserver/sqlserver.go [hunk 24]
```diff
@@ -423,6 +464,96 @@ func (s *SQLServer) Init() error {
...
+	spt, err := adal.NewServicePrincipalTokenFromMSI(msiEndpoint, sqlAzureResourceID)
+	if err != nil {
+		return nil, err
+	}
+
+	// ensure token is fresh
+	err = spt.EnsureFresh()
+	if err != nil {
+		return nil, err
+	}
```
Comment (srebhan) [inline plugins/inputs/sqlserver/sqlserver.go:541]: You might want to use
```suggestion
	if err := spt.EnsureFresh(); err != nil {
		return nil, err
	}
```
here.

**Review by srebhan — CHANGES_REQUESTED** [review]
I have some comments in the code. The most sever is the potential deadlock when an error occurs in `getTokenProvider()`. Please take a look and fix the leaking lock.

File: plugins/inputs/sqlserver/sqlserver.go (hunk 17 above)
Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserver.go]: When I initially wrote this I did have a reason, however now I don't think it makes sense to keep it out of the struct. Moved them to the struct.

File: plugins/inputs/sqlserver/sqlserver.go (hunk 18 above)
Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserver.go]: done

File: plugins/inputs/sqlserver/sqlserver.go (hunk 19 above)
Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserver.go]: Yes, I noticed it after merging the latest changes (for connection pooling). I was already adding the error to accumulator and this printf is not required, refactored it.

File: plugins/inputs/sqlserver/sqlserver.go (hunk 20 above)
Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserver.go]: refactored it

File: plugins/inputs/sqlserver/sqlserver.go (hunk 21 above)
Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserver.go]: Thank you for catching this, fixed it

File: plugins/inputs/sqlserver/sqlserver.go (hunk 23 above)
Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserver.go]: Yes, it a fixed resource url, moved it with other const declarations

**Comment by avinash-nigam** [comment]
> I have some comments in the code. The most sever is the potential deadlock when an error occurs in `getTokenProvider()`. Please take a look and fix the leaking lock.

Thank you @srebhan for the review, I've addressed your comments.

File: plugins/inputs/sqlserver/sqlserver.go (hunk 22 above)
Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserver.go]: This was added while looking for different options of caching the token (for example, disk based), so would be useful if we use a different caching mechanism. I'd prefer to keep it structured this way, unless you have a strong opinion on it.

Comment (srebhan) [reply plugins/inputs/sqlserver/sqlserver.go]: Fine with me, maybe you could add a comment about this reasoning to the code?

Comment (srebhan) [reply plugins/inputs/sqlserver/sqlserver.go]: Btw: Error messages should start lowercase... ;-P

**Review by srebhan — COMMENTED** [review]
Only some small (cosmetic) things left. Looks good so far.

File: plugins/inputs/sqlserver/sqlserver.go (hunk 22 above)
Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserver.go]: Added a comment to the method and the error message with lowercase was already fixed in the previous commit (above screen shot is "outdated" :-) )

File: plugins/inputs/sqlserver/sqlserver.go [hunk 25]
```diff
@@ -296,29 +296,29 @@ func (s *SQLServer) Start(acc telegraf.Accumulator) error {
...
+				continue
 			}
 		} else {
 			// otherwise assume AAD Auth with system-assigned managed identity (MSI)
 
 			// AAD Auth is only supported for Azure SQL Database or Azure SQL Managed Instance
 			if s.DatabaseType == "SQLServer" {
-				return fmt.Errorf("Database connection failed : AAD auth is not supported for SQL VM i.e. DatabaseType=SQLServer")
+				er := errors.New("database connection failed : AAD auth is not supported for SQL VM i.e. DatabaseType=SQLServer")
+				acc.AddError(er)
```
[the remark this answers is not in the document]
Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserver.go]: changed to err

**Comment by avinash-nigam** [comment]
> Only some small (cosmetic) things left. Looks good so far.

Thank you @srebhan for quick review, I've addressed your comments.

**Review by masree — APPROVED** [review]

File: plugins/inputs/sqlserver/sqlserver.go [hunk 26]
```diff
@@ -423,6 +463,94 @@ func (s *SQLServer) Init() error {
 	return nil
 }
 
+// Get Token Provider by loading cached token or refreshed token
+func (s *SQLServer) getTokenProvider() (func() (string, error), error) {
```
Comment (masree) [inline plugins/inputs/sqlserver/sqlserver.go:467]: Can we move token related methods to a different class? It would keep the main class short and simple

Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserver.go:467]: I feel the methods are not reusable as such to warrant a refactor to another class. And after moving the cache variable and it's lock into the SQLServer struct, those methods are more bound and separating the methods to another class won't add value. I'd prefer to keep them here for now unless we change the caching mechanism to a more reusable structure.

**Review by srebhan — APPROVED** [review]
Looks good to me.

**Comment by sspaink** [comment]
> How have you been testing it? is there a testing plan you would like to link for us to take a look?
> Example -
> Scenario1: Telegraf installed on VM and with connection strings using sqlauth only
> Scenario2: Telegraf installed on VM with connection strings using sqlauth and aadMSI (multiple connection strings)
> and so on

@avinash-nigam I was also wondering about how you tested this? There aren't any unit tests covering this new auth method, what are your thoughts on including tests?

**Comment by avinash-nigam** [comment]
> > How have you been testing it? is there a testing plan you would like to link for us to take a look?
> > Example -
> > Scenario1: Telegraf installed on VM and with connection strings using sqlauth only
> > Scenario2: Telegraf installed on VM with connection strings using sqlauth and aadMSI (multiple connection strings)
> > and so on
> 
> @avinash-nigam I was also wondering about how you tested this? There aren't any unit tests covering this new auth method, what are your thoughts on including tests?

@sspaink - This has been manually tested for databases of following variants individually and a combination of multiple variants by deploying the bits on an Ubuntu VM and settings connection strings to use SqlAuth and AADAuth (both as supported cases and not supported cases) - 

Azure SQL Database - S1, S2 (AAD supported)
Azure SQL Elastic Pool Database - S0, S2 (AAD supported)
SQL VM - 2012, 2014, 2016, 2019 (AAD not supported)
SQL AG (HADR) - 2019 (AAD not supported)

Currently this plugin does not support integration testing, so there are no tests for authentication yet. Once we build a mechanism to write integration tests, then we would be able to add tests for the same.

**Comment by cerilewis** [comment]
This change appears to break Windows Integrated authentication where no User ID or Password is given in the connection string. If I run telegraf version 1.19.0-rc0 with configuration that works with 1.18.3 I get the following error:
`2021-06-14T09:35:28Z E! [inputs.sqlserver] Error in plugin: database connection failed : AAD auth is not supported for SQL VM i.e. DatabaseType=SQLServer`

This appears to come from the check for Password attribute being used to determine whether to user SQL Server authentication or AAD.

**Comment by srebhan** [comment]
@cerilewis thanks for reporting this problem! Can you please open an issue and link to this PR. I fear that we need to add an option to specify which auth-scheme to use as the magic here will likely not be able to detect it... :-(

**Comment by cerilewis** [comment]
@srebhan I have created the issue  #9362 but either can't or don't know how to link this PR to it correctly.

**Comment by srebhan** [comment]
You already did by mentioning this PR in the issue. :-) Thank you for reporting.

**Comment by srebhan** [comment]
@arkapravasinha please do not post support questions to old PRs! Nobody will notice them besides me. ;-) Better use the forum or Slack!

This being said, how about reading the [docs ](https://github.com/influxdata/telegraf/blob/master/plugins/inputs/sqlserver/README.md) and use `auth_method = "aad"`?
